src/chainstate/stacks/boot/lockup.clar

```diff
@@ -0,0 +1,148 @@
+;; The .lockup boot contract
+;; Error codes
+(define-constant ERR-UNLOCK-UNREACHABLE 255)
+
+;; Map for all locked STX that will later unlock.
+;; Entries in this map exist from Stacks 1.0.  These are tokens that were purchased
+;; in one or more of the token sales from Blockstack PBC and its subsidiaries.
+;; Some of these tokens are illiquid until certain block heights are reached.
+;; This map encodes these block heights, and binds them to their owner and unlock
+;; amount.
+;; 
+;; This map is populated in the boot block from data pulled from the
+;; Stacks 1.0 genesis block.  No smart contract can add to this map.
+;; Rows are deleted from this map once they're processed.
+;;
+;; The map here is structured like an array.  Each block height is coupled with
+;; an index ranging from 0 to the number of principals who unlock tokens in this
+;; block height, and that number of principals is stored in the internal-locked-stx-len
+;; The miner and block-processing code will use this map and the internal-locked-stx-len
+;; map to iterate through each Stacks block's rows and process them via the
+;; private (internal-unlock-stx) method.  The node's Clarity VM will take care of
+;; calling the method; no smart contracts will ever interact with this state
+;; directly.
```

Comment:
If what we need is an array, why using a map instead of a list?

Reply:
I was debating this as well.  I went with implementing an append-only array on top of a data-map because I didn't want users to have to pay for loading/storing a gigantic list each time they registered a PoX address (and moreover, I don't want the block's I/O budget to get expended on just PoX lockups, which would penalize _everyone_).  The downside is that the Stacks node has to do more work to load the reward set when entering the next reward cycle, but I don't think this cost should be borne by users.

Reply:
Oh wait nm this is the `lockup.clar` code, not the `pox.clar` code.  The code in `lockup.clar` is pretty open to debate; in fact, I'm happy to remove it altogether.  I thought at one point when writing this PR that I'd need to test `total-liquid-ustx` in the face of tokens unlocking, but this wasn't necessary (the act of mined tokens maturing has the same effect I needed).